x/oracle/types/prophecy.go

```diff
@@ -0,0 +1,143 @@
+	claimValidators, err := json.Marshal(prophecy.ClaimValidators)
+	if err != nil {
+		return DBProphecy{}, err
+	}
+
+	validatorClaims, err := json.Marshal(prophecy.ValidatorClaims)
+	if err != nil {
+		return DBProphecy{}, err
+	}
+
+	return DBProphecy{
+		ID:              prophecy.ID,
+		Status:          prophecy.Status,
+		ClaimValidators: claimValidators,
+		ValidatorClaims: validatorClaims,
+	}, nil
+}
+
+// DeserializeFromDB deserializes a DBProphecy into a prophecy
+func (dbProphecy DBProphecy) DeserializeFromDB() (Prophecy, error) {
+	var claimValidators map[string][]sdk.ValAddress
+	err := json.Unmarshal(dbProphecy.ClaimValidators, &claimValidators)
+	if err != nil {
+		return Prophecy{}, err
+	}
+
+	var validatorClaims map[string]string
+	err = json.Unmarshal(dbProphecy.ValidatorClaims, &validatorClaims)
+	if err != nil {
+		return Prophecy{}, err
+	}
+
+	return Prophecy{
+		ID:              dbProphecy.ID,
+		Status:          dbProphecy.Status,
+		ClaimValidators: claimValidators,
+		ValidatorClaims: validatorClaims,
+	}, nil
+}
+
+// AddClaim adds a given claim to this prophecy
+func (prophecy Prophecy) AddClaim(validator sdk.ValAddress, claim string) {
+	claimValidators := prophecy.ClaimValidators[claim]
+	prophecy.ClaimValidators[claim] = append(claimValidators, validator)
+
+	validatorBech32 := validator.String()
+	prophecy.ValidatorClaims[validatorBech32] = claim
+}
+
+func (prophecy Prophecy) FindHighestClaim(ctx sdk.Context, stakeKeeper staking.Keeper) (string, int64, int64) {
+	validators := stakeKeeper.GetBondedValidatorsByPower(ctx)
+	//Index the validators by address for looking when scanning through claims
+	validatorsByAddress := make(map[string]staking.Validator)
+	for _, validator := range validators {
+		validatorsByAddress[validator.OperatorAddress.String()] = validator
+	}
+
+	totalClaimsPower := int64(0)
+	highestClaimPower := int64(-1)
+	highestClaim := ""
+	for claim, validators := range prophecy.ClaimValidators {
+		claimPower := int64(0)
+		for _, validator := range validators {
+			validatorPower := validatorsByAddress[validator.String()].GetTendermintPower()
+			claimPower += validatorPower
+		}
+		totalClaimsPower += claimPower
+		if claimPower > highestClaimPower {
+			highestClaimPower = claimPower
+			highestClaim = claim
+		}
+	}
+	return highestClaim, highestClaimPower, totalClaimsPower
+}
+
+// NewProphecy returns a new Prophecy, initialized in pending status with an initial claim
+func NewProphecy(id string) Prophecy {
+	return Prophecy{
+		ID:              id,
+		Status:          NewStatus(PendingStatusText, ""),
+		ClaimValidators: make(map[string][]sdk.ValAddress),
+		ValidatorClaims: make(map[string]string),
+	}
+}
+
+// NewEmptyProphecy returns a blank prophecy, used with errors
+func NewEmptyProphecy() Prophecy {
+	return NewProphecy("")
+}
+
+// Status is a struct that contains the status of a given prophecy
+type Status struct {
+	Text       string `json:"text"`

```

Comment:
I think the `status.Text` should be enum. Think this can be done using `iota` in go

Reply:
done in next push